File: src/lib/api/env.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  EnvConflict,
  BackupInfo,
  EnvRestoreDiff,
} from "@/types/env";

/**
 * 环境变量管理 API
 */

/**
 * 检查指定应用的环境变量冲突
 * @param appType 应用类型（与 env_checker 后端的 get_keywords_for_app 对齐）
 * @returns 环境变量冲突列表
 */
export async function checkEnvConflicts(
  appType: string,
): Promise<EnvConflict[]> {
  return invoke<EnvConflict[]>("check_env_conflicts", { app: appType });
}

/**
 * 删除指定的环境变量 (会自动备份)
 * @param conflicts 要删除的环境变量冲突列表
 * @returns 备份信息
 */
export async function deleteEnvVars(
  conflicts: EnvConflict[],
): Promise<BackupInfo> {
  return invoke<BackupInfo>("delete_env_vars", { conflicts });
}

/**
 * 判断变量是否被 OGG 里配置的供应商正在使用
 * @returns 变量名(大写) -> 引用它的 "app · 供应商名" 列表
 */
export async function envVarsInUse(
  varNames: string[],
): Promise<Record<string, string[]>> {
  return invoke<Record<string, string[]>>("env_vars_in_use", { varNames });
}

/**
 * 列出可恢复的环境变量备份（最新在前）
 */
export async function listEnvBackups(): Promise<
  { backupPath: string; timestamp: string; varNames: string[] }[]
> {
  return invoke("list_env_backups");
}

/**
 * 读取单个备份（只读，不写入任何东西）
 */
export async function readEnvBackup(
  backupPath: string,
): Promise<BackupInfo> {
  return invoke<BackupInfo>("read_env_backup", { backupPath });
}

/**
 * 恢复前对比：给出每条变量的备份值 / 当前值 / 是否会被覆盖
 */
export async function diffEnvBackup(backupPath: string): Promise<EnvRestoreDiff> {
  return invoke<EnvRestoreDiff>("diff_env_backup", { backupPath });
}

/**
 * 从备份文件恢复环境变量
 * @param backupPath 备份文件路径
 */
export async function restoreEnvBackup(backupPath: string): Promise<void> {
  return invoke<void>("restore_env_backup", { backupPath });
}

/**
 * 检查所有应用的环境变量冲突
 * @returns 按应用类型分组的环境变量冲突
 *
 * app 列表 = OGG 实际管理环境变量语义的应用。历史上这里是写死的
 * claude/codex/gemini/grokbuild 四项——前三个来自已裁撤的运行时代码，
 * 只有它们的 env_checker 关键词还在。保留 claude/codex/gemini 是为了
 * 兼容用户手工遗留的 ANTHROPIC_/OPENAI_/GEMINI_ 变量，但加入 antigravity
 * （复用 GEMINI_ 关键词，受管键已在后端排除）。
 */
export async function checkAllEnvConflicts(): Promise<
  Record<string, EnvConflict[]>
> {
  const apps = ["claude", "codex", "gemini", "grokbuild", "antigravity"];
  const results: Record<string, EnvConflict[]> = {};

  await Promise.all(
    apps.map(async (app) => {
      try {
        results[app] = await checkEnvConflicts(app);
      } catch (error) {
        console.error(`检查 ${app} 环境变量失败:`, error);
        results[app] = [];
      }
    }),
  );

  return results;
}

File: src-tauri/src/commands/env.rs
use crate::services::env_checker::{check_env_conflicts as check_conflicts, EnvConflict};
use crate::services::env_manager::{
    delete_env_vars as delete_vars, list_backups, read_backup, restore_from_backup, BackupInfo,
    BackupSummary, EnvRestoreDiff,
};

/// 判断一组环境变量是否被 OGG 里配置的供应商**正在使用**。
///
/// 删除前的关键信号：一个变量即使不是 OGG 受管键，也可能正是某供应商条目里存的
/// 凭据/端点。此时删掉它，供应商卡片还在但会失效——用户无从察觉。
///
/// 判定方式：逐 app 读供应商表，凡 `settingsConfig.env` 里存在同名键且值与待删值
/// 相同的，计为一次引用。注意只比对「当前生效槽位」（持久环境变量语义下同一时刻
/// 只有一个值在用），因此不做跨值匹配。
#[tauri::command]
pub fn env_vars_in_use(
    state: tauri::State<'_, crate::store::AppState>,
    var_names: Vec<String>,
) -> Result<std::collections::HashMap<String, Vec<String>>, String> {
    use crate::services::env_manager::EnvVarUsage;
    let mut out: std::collections::HashMap<String, Vec<String>> =
        std::collections::HashMap::new();
    let wanted: Vec<String> = var_names.iter().map(|n| n.to_uppercase()).collect();
    if wanted.is_empty() {
        return Ok(out);
    }

    for app in ["grokbuild", "antigravity", "omp"] {
        let providers = match state.db.get_all_providers(app) {
            Ok(p) => p,
            Err(_) => continue,
        };
        for provider in providers.values() {
            let env = provider
                .settings_config
                .get("env")
                .and_then(serde_json::Value::as_object);
            let Some(env) = env else { continue };
            for (key, value) in env {
                let Some(value) = value.as_str() else { continue };
                if value.trim().is_empty() {
                    continue;
                }
                let upper = key.to_uppercase();
                // 只关心待删除列表里的键
                if !wanted.iter().any(|w| *w == upper) {
                    continue;
                }
                out.entry(upper.clone())
                    .or_default()
                    .push(EnvVarUsage::label(app, &provider.name));
            }
        }
    }
    Ok(out)
}

/// Check environment variable conflicts for a specific app
#[tauri::command]
pub fn check_env_conflicts(app: String) -> Result<Vec<EnvConflict>, String> {
    check_conflicts(&app)
}

/// Delete environment variables with backup
#[tauri::command]
pub fn delete_env_vars(conflicts: Vec<EnvConflict>) -> Result<BackupInfo, String> {
    delete_vars(conflicts)
}

/// List recoverable environment variable backups (newest first)
#[tauri::command]
pub fn list_env_backups() -> Result<Vec<BackupSummary>, String> {
    list_backups()
}

/// Read one backup file for a "before restoring" review (no writes)
#[tauri::command]
pub fn read_env_backup(backup_path: String) -> Result<BackupInfo, String> {
    read_backup(&backup_path)
}

/// Compare a backup against the current machine state: per variable, whether it is
/// currently set, whether the value differs, and whether an OGG-managed provider
/// would be affected. Restoring overwrites unconditionally — this exists so the UI
/// can show that **before** anything is written.
#[tauri::command]
pub fn diff_env_backup(backup_path: String) -> Result<EnvRestoreDiff, String> {
    crate::services::env_manager::diff_backup(&backup_path)
}

/// Restore environment variables from backup file
#[tauri::command]
pub fn restore_env_backup(backup_path: String) -> Result<(), String> {
    restore_from_backup(backup_path)
}
